File: year2020/day03/java/Solution.java
import java.nio.file.*;
import java.util.*;

public class Solution {
    private static List<String> TREES;

    public static void main(String[] args) { 
        try {
            setup(Files.readAllLines(Paths.get("../puzzle.txt")));
            System.out.println(part1());
            System.out.println(part2());
        } catch(Exception e) {
            e.printStackTrace();
        }
    }
    
	public static int part1() {
        return checkTrees(3, null);
    }
    
    public static long part2() {
        int[] steps = {1, 3, 5, 7};
        return Arrays.stream(steps).mapToLong(step -> checkTrees(step, null)).reduce(1, (acc, result) -> acc * result) * checkTrees(1, 1);      
    }

    public static void setup(List<String> input) {
        TREES = input;
    }

    private static int checkTrees(int step, Integer skip) {
        int index = 0;
        int counter = 0;
        for (int i = 0; i < TREES.size(); i++) {
            if (TREES.get(i).charAt(index) == '#') {
                counter++;
            }
            index = (index + step) % TREES.get(i).length();
            if (skip != null) {
                i += skip;
            }
        }
        return counter;
    }
}

File: year2020/day03/python/solution.py
def main():
    print(part_1())
    print(part_2())


def part_1():
    return check_trees(input, 3)


def part_2():
    product = 1
    for step in [1, 3, 5, 7]:
        product *= check_trees(input, step)

    return product * check_trees(input, 1, 1)


def setup(path):
    global input
    
    with open(path, 'r') as file:
        input = [row.rstrip() for row in file]


def check_trees(input, step, skip=None):
    index, counter, i = 0, 0, 0
    while i < len(input):
        if input[i][index] == '#':
            counter += 1

        index = (index + step) % len(input[i])
        i = i + skip if skip is not None else i
        i += 1

    return counter
 

if __name__ == '__main__':
    setup('../puzzle.txt')
    main()